### ISSUE TITLE
Multiple coronagraphy roll angles not being combined properly in CORON3

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#5502]
_Issue [JP-1790](https://jira.stsci.edu/browse/JP-1790) was created on JIRA by [Ben Sunnquist](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bsunnquist):_

NIRCam coronagraphy target images with 2 different rolls angles aren't combined properly in the CORON3 pipeline; the final combined psf-subtracted image shows both of these images on different parts of the sky (see attached), even though they are the same target and only differ in their roll angle (10 degrees).

It looks like the pipeline correctly accounts for the different roll angles (the images are rotated differently in the final combined image), but then puts them onto different parts of the sky. Besides the roll angle, most WCS header keywords are identical between these two target images (same TARG_RA/DEC, RA/DEC_REF, V2/V3_REF, CRVAL1/2), so we expected them to be combined on top of eachother. The PC matrix and RA/DEC_V1 values are different between the two images, but it looks like the pipeline doesn't use these for this, and the results don't change if you make them equal, so I don't think that's the issue (though let me know if there are other keywords that may be affecting things to be sure it's not a simulation issue).

This issue first becomes apparent in the image2 assign_wcs step; the populated S_REGION keyword here is on different parts of the sky for these 2 target images.

The provided data directory contains all of the inputs/outputs used in the various stages of the pipeline (detector1 -> image2 -> coron3). This simulated dataset consists of 9 small-grid dithered reference PSFs and the 2 target PSFs with different roll angles.

**Comment by stscijgbot-jp** [comment]
_Comment by [Julien Girard](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jgirard) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=499199#comment-499199):_

So, if I understand correctly the Coron3 align_refs step also registers the science images
<https://jwst-pipeline.readthedocs.io/en/latest/jwst/align_refs/description.html#d-calibrated-images>

But in this case, they are too far apart from image2 assign_wcs step.

We have to keep in mind that in general, the masks won't be located at the center of the subarrays and so I would think that in case the WCS isn't present or not dealt with correctly by image2, the Coron3 align_refs step should be robust enough to align the frames:
* in the general case when stars/masks aren't in the center (especially true for the BAR masks).
* in the event the images are completely separated from WCS issue (or absence thereof).
Thanks [Ben Sunnquist](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bsunnquist) for looking into this. I think we should work on these Coron3 workarounds while image2 is being checked. Probably we need to simulate "less ideal" data to make sure we can test robustness.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=499214#comment-499214):_

[Julien Girard](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jgirard) No, the align_refs step only aligns PSF images to the science target images, so that the KLIP operation (PSF subtraction) can be performed properly. The science target images do not undergo any alignment to themselves. Right now the alignment is completely based on the WCS pointing information for each target image, which is used as input to the resample (drizzle) step to combine the PSF-subtracted target images. Right now there's no equivalent of the tweakreg step that's applied anywhere in that flow in order to do a self registration of the images before combining/drizzling. So if the WCS info is off, the combined image will have problems.

**Comment by stscijgbot-jp** [comment]
_Comment by [Julien Girard](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jgirard) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=499229#comment-499229):_

Thanks [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) for this clarification. I see. Do you think then that in this case, it's the WCS in our simulations (headers) that is not adequate or the image2 assign_wcs step that does not treat properly the coronagraphic images?

I have another question: If roll 1 and roll 2 happen to be slightly misaligned, I suppose there will be one set of PSF images registered to roll 1 and another set registered to roll 2, right?
Or by definition, this is impossible because roll 1 and roll 2 WILL have the same WCS?

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=499308#comment-499308):_

It could be due to something not being quite right or complete in the WCS info in the simulated images, we'll need to check that as we investigate.

Regarding the registration of PSF images, yes, the registration is done on an exposure-by-exposure and even integration-by-integration basis. So for example if you have 1 PSF exposure consisting of 10 integrations (in a calints file) and 2 target exposures (one at roll 1 and the other at roll 2) each again consisting of 10 integrations, the image in each of the 10 integrations of the PSF exposure is aligned separately to each of the 20 integrations of the 2 target exposures. This alignment (using the routines in <https://github.com/spacetelescope/jwst/blob/master/jwst/coron/imageregistration.py)> does not use an WCS info for any of the images. It is based purely on FFT techniques for cross-correlating the actual source signals in each image. That resulting stack of aligned PSF and target images is what KLIP then works on, yielding a PSF-subtracted result for each of the total of 20 target integrations individually (10 from roll 1 and 10 from roll 2). Those PSF-subtracted target images inherit the original WCS of each target exposure, which is then used to combine the 20 PSF-subtracted images into a single result. If the WCS of the original exposures correctly includes the differences in roll angles, they should all be aligned to the same roll when combining.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=504116#comment-504116):_

After discussing this with [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) and James Davies and with the help of [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse)'s explanation of the processing of these images, I believe that the most likely cause of observed mis-alignment is due to how data were simulated.

The roll angle is with regard to the telescope axis and the chip being offset from the axis, one would expect the detector to end up, after a roll, in a different (offset) position when looked from the plane of the first image. If one wants to have the same star to appear at approximately the same position in the detector/image but be rotated compared to the first observation then RA_REF and DEC_REF on 'wcsinfo' of the two observations should be different. That is, after a telescope roll, it's pointing should be re-adjusted. However, both science images at different roll angles have identical RA_REF and DEC_REF which results in a large shift (because the roll is large).

**Comment by stscijgbot-jp** [comment]
_Comment by [Ben Sunnquist](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bsunnquist) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=504940#comment-504940):_

Ah that makes sense - thank you for looking into that Mihai! And apologies for the simulation issue.

I'll recreate these images with the corrected RA/DEC_REF and re-run. I'll update this ticket once that is complete and everything checks out.

**Comment by nden** [comment]
The issue was resolved and the ticket withdrawn.

**Comment by stscijgbot-jp** [comment]
_Comment by [Ben Sunnquist](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bsunnquist) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=507503#comment-507503):_

Yep, after updating the RA/DEC_REF (as well as another header issue I found that resulted in the wrong distortion reffile applied), everything looks good (see new attached image), so I think this issue can be closed.

I've updated the simulation notebook to account for both of these issues in the future. Thanks for the help and apologies for the simulation issue!

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=499214#comment-499214):_

[Julien Girard](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jgirard) No, the align_refs step only aligns PSF images to the science target images, so that the KLIP operation (PSF subtraction) can be performed properly. The science target images do not undergo any alignment to themselves. Right now the alignment is completely based on the WCS pointing information for each target image, which is used as input to the resample (drizzle) step to combine the PSF-subtracted target images. Right now there's no equivalent of the tweakreg step that's applied anywhere in that flow in order to do a self registration of the images before combining/drizzling. So if the WCS info is off, the combined image will have problems.

**Comment by stscijgbot-jp** [comment]
_Comment by [Julien Girard](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jgirard) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=499229#comment-499229):_

Thanks [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) for this clarification. I see. Do you think then that in this case, it's the WCS in our simulations (headers) that is not adequate or the image2 assign_wcs step that does not treat properly the coronagraphic images?

I have another question: If roll 1 and roll 2 happen to be slightly misaligned, I suppose there will be one set of PSF images registered to roll 1 and another set registered to roll 2, right?
Or by definition, this is impossible because roll 1 and roll 2 WILL have the same WCS?

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=499308#comment-499308):_

It could be due to something not being quite right or complete in the WCS info in the simulated images, we'll need to check that as we investigate.

Regarding the registration of PSF images, yes, the registration is done on an exposure-by-exposure and even integration-by-integration basis. So for example if you have 1 PSF exposure consisting of 10 integrations (in a calints file) and 2 target exposures (one at roll 1 and the other at roll 2) each again consisting of 10 integrations, the image in each of the 10 integrations of the PSF exposure is aligned separately to each of the 20 integrations of the 2 target exposures. This alignment (using the routines in <https://github.com/spacetelescope/jwst/blob/master/jwst/coron/imageregistration.py)> does not use an WCS info for any of the images. It is based purely on FFT techniques for cross-correlating the actual source signals in each image. That resulting stack of aligned PSF and target images is what KLIP then works on, yielding a PSF-subtracted result for each of the total of 20 target integrations individually (10 from roll 1 and 10 from roll 2). Those PSF-subtracted target images inherit the original WCS of each target exposure, which is then used to combine the 20 PSF-subtracted images into a single result. If the WCS of the original exposures correctly includes the differences in roll angles, they should all be aligned to the same roll when combining.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=504116#comment-504116):_

After discussing this with [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) and James Davies and with the help of [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse)'s explanation of the processing of these images, I believe that the most likely cause of observed mis-alignment is due to how data were simulated.

The roll angle is with regard to the telescope axis and the chip being offset from the axis, one would expect the detector to end up, after a roll, in a different (offset) position when looked from the plane of the first image. If one wants to have the same star to appear at approximately the same position in the detector/image but be rotated compared to the first observation then RA_REF and DEC_REF on 'wcsinfo' of the two observations should be different. That is, after a telescope roll, it's pointing should be re-adjusted. However, both science images at different roll angles have identical RA_REF and DEC_REF which results in a large shift (because the roll is large).

**Comment by stscijgbot-jp** [comment]
_Comment by [Ben Sunnquist](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bsunnquist) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=504940#comment-504940):_

Ah that makes sense - thank you for looking into that Mihai! And apologies for the simulation issue.

I'll recreate these images with the corrected RA/DEC_REF and re-run. I'll update this ticket once that is complete and everything checks out.

**Comment by stscijgbot-jp** [comment]
_Comment by [Ben Sunnquist](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bsunnquist) on [JIRA](https://jira.stsci.edu/browse/JP-1790?focusedCommentId=507503#comment-507503):_

Yep, after updating the RA/DEC_REF (as well as another header issue I found that resulted in the wrong distortion reffile applied), everything looks good (see new attached image), so I think this issue can be closed.

I've updated the simulation notebook to account for both of these issues in the future. Thanks for the help and apologies for the simulation issue!
